src/oidc.rs

```diff
@@ -13,6 +13,35 @@ use time::now_utc;
 use url::percent_encoding::{utf8_percent_encode, QUERY_ENCODE_SET};
 
 
+/// Macro used to extract a bunch of parameters from a JSON Value.
+///
+/// This macro will return from the contained method with a BrokerResult when
+/// it fails to parse one of the params. The error will contain the description
+/// string from the second parameter.
+///
+/// ```
+/// extract_json_fields!(value, "example document", {
+///     foo = as_str("foo"),
+///     bar = as_i64("BAR"),
+/// });
+/// ```
+macro_rules! extract_json_fields {
```

Comment:
Man, I don't know about this, it again feels very magic. Can get most of the value with less magic?